### ISSUE TITLE
Distribute darknet node addresses via DNS seeds using AAAA records

### ISSUE THREAD

**Issue by virtu** [bitcoin/bitcoin#31062]
### Please describe the feature you'd like to see added.

Right now, Bitcoin Core can only receive IPv4 and IPv6 node addresses from DNS seeds. Adding support for darknet addresses would bring the advantages of using DNS seeds for node discovery to darknet nodes.

### Describe the solution you'd like

Before TorV2 addresses were deprecated, [Bitcoin Core](https://github.com/bitcoin/bitcoin/blob/5fe6878b5f7c1c97b5c8ae04be9592be73e840c1/src/netaddress.h#L64-L69) and seeds using [sipa's seeder](https://github.com/sipa/bitcoin-seeder/blob/ff482e465ff84ea6fa276d858ccb7ef32e3355d3/netbase.cpp#L535-L543) (and maybe others) supported distributing such addresses by encoding them in AAAA records: Since TorV2 addresses used only 80-bit hashes, they neatly fitted into the 128-bit data field of an AAAA record, leaving plenty of room for a 48bit prefix (from the not publicly routable fc00::/7 subnet) to signal the custom encoding.

TorV3 and I2p addresses use 256-bits of data, making them too large for single AAAA records. However, such addresses can be serialized using a trimmed BIP155 format (just the net id and address), concatenated, and broken into chunks that fit into AAAA record data fields.

In addition to a prefix to signal the custom encoding, correct deserialization requires a sequence number since public resolvers may reorder records.

To get a rough estimate for the numbers, consider the following: DNS messages should be at most 512B. The header has a fixed size of 12B. The question section has variable size because it contains the query domain. 30 chars should be sufficient for most seeds, which implies a size of 36B for the question section (32B for the 30 chars, plus 2B each for query type and class). This leaves 464B for the record section. Since AAAA records are 28B each (2B each for name pointer, type, class, length, 4B for TTL, and 128b/16B for data), a DNS reply for regular-length domains can at most contain 16 AAAA records. Hence four bits are required to encode sequence numbers.

The publicly unroutable fc00::/7 prefix used in the past uses 7 bits. Extending these by e.g. five more bits for versioning to accommodate for future updates results in a signaling-prefix size of 12 bits, leaving 112 bit (or 14B) for payload:

```
 0         11 12       15 16                                                       127
+------------+----------+------------------------------------------------------------+
|  Prefix    |   Seq#   |                         Payload                            |
|  (12 bits) | (4 bits) |                        (112 bits)                          |
+------------+----------+------------------------------------------------------------+
```

At 16 AAAA records per DNS message, 14B of payload per record and 33B per TorV3/I2P address (net id plus 256-bit data), six TorV3/I2P addresses can be encoded per message.

CJDNS addresses, which are IPv6 addresses, could be encoded as well; at 17B per address (net id plus 128-bit data), 13 addresses could be encoded per message.

Demo implementation: https://github.com/virtu/darkseed
Demo DNS seed: `dnsseed.21.ninja`

### Describe any alternatives you've considered

Originally, I considered using DNS NULL records, because their data fields allow for an arbitrary amount of binary data, resulting in a much better encoding efficiency than ~50% offered by the proposed AAAA encoding (14B of overhead-including payload per 28B record).

This approach, however, has the following shortcomings (as pointed out by [sipa in the forum](https://delvingbitcoin.org/t/hardcoded-seeds-dns-seeds-and-darknet-nodes/1123/6)):
- Depend on external DNS library or extra code to create DNS request with NULL query type
- Public recursive resolvers might not cache DNS NULL records

**Comment by 1440000bytes** [comment]
Concept NACK

Onion and i2p nodes not dependent on DNS seeds is a feature. It's better if only IPv4 and IPv6 nodes use it.

Maybe we could experiment with [GNS](https://www.gnunet.org/en/gns.html)? Or use this approach for IPv4 nodes with non default port? https://github.com/bitcoin/bitcoin/issues/30900

**Comment by laanwj** [comment]
Concept NACK

It's a clever trick, but we've considered this kind of thing before and it's a really a deep rabbit hole.

From what i heard it's not generally a good idea to stash different data in AAAA records. Caching DNS servers can have checks to assume the addresses are routable and valid. Additionally, there's also no guarantee that clients get all the addresses, so splitting the data is a bit brittle.

This would also break the assumption that a random address returned from the DNS seed is connectable (because they might get a fragment instead).  Bitcoin core peers fully behind Tor rely on this behavior.

Also wonder, like @1440000bytes, what would be the motivation to have I2P and Tor addresses returned by DNS seeds? Mind that clients behind neither can do DNS queries (well, Tor has a [SOCKS extension](https://spec.torproject.org/socks-extensions.html#name-lookup) to do a DNS query, but it cannot return more than one result), and querying through clearnet through ISP's DNS servers is a leak.

One could return arbitrary data in TXT records but they're much harder to query from C++ code, and suffer from the same above problem.

**Comment by vasild** [comment]
I agree with the reasoning above. I am still Concept ACK if we can have the clients get all addresses from the response, without some addresses being omitted. I see no downsides if separate DNS seeds are used for the new encoding, leaving the existent ones as they are currently.

Just to clarify - Tor only nodes do not use DNS seeds and the intention here is to keep it that way. The motivation is to help multi-homed nodes that use clearnet and Tor. They are currently only getting clearnet addresses from the DNS seeds.

To include ports in the DNS seeds responses is a separate motivation for this.

**Comment by willcl-ark** [comment]
Might it not make more sense to avoid hacking DNS responses for this, or using DNS at all, and simply include some native Tor (/I2P/CJDNS) addresses in the `vSeeds` chainparams list? These hard-coded addresses could still point to a DNS seeder which can respond to addr_fetch requests from its hidden service (or I2P equivalent) with addresses of that type only. @achow101 [dnsseedrs](https://github.com/achow101/dnsseedrs/tree/main) already has onion, I2P and CJDNS results available to query, for example.

This avoids any requirement to use DNS over Tor (which may not be available anyway), and, whilst we may lose some of the other benefits of using DNS (like caching of addrs in the DNS itself), the marginal privacy benefits of requesting addrs via DNS seeds, which someone using a recursive resolver might get, are already included as part of connecting via an anonymity network.

**Comment by pinheadmz** [comment]
Do we need to hack DNS records for this? We have [BIP155](https://github.com/bitcoin/bips/blob/master/bip-0155.mediawiki) addrv2 messages to relay onion peers, all you need is one bitcoin peer to collect those.

The hard coded seeds list also provides onion, CJDNS, and i2p addresses.
